src/python/pants/backend/python/tasks/setup_py.py

```diff
@@ -79,21 +95,87 @@ def _setup_command(self):
     return self.__setup_command
 
 
-class SetupPyInvocationEnvironment(datatype(['joined_path'])):
+class SetupPyNativeTools(datatype([
+    ('c_compiler', CCompiler),
+    ('cpp_compiler', CppCompiler),
+    ('linker', Linker),
+])):
+  """The native tools needed for a setup.py invocation.
+
+  This class exists because `SetupPyExecutionEnvironment` is created manually, one per target.
+  """
+
+
+@rule(SetupPyNativeTools, [Select(CCompiler), Select(CppCompiler), Select(Linker)])
+def get_setup_py_native_tools(c_compiler, cpp_compiler, linker):
+  yield SetupPyNativeTools(c_compiler=c_compiler, cpp_compiler=cpp_compiler, linker=linker)
+
+
+class SetupRequiresSiteDir(datatype(['site_dir'])): pass
+
+
+# TODO: This could be formulated as an @rule if targets and `PythonInterpreter` are made available
+# to the v2 engine.
+def ensure_setup_requires_site_dir(reqs_to_resolve, interpreter, site_dir,
+                                   platforms=None):
+  if not reqs_to_resolve:
+    return None
+
+  setup_requires_dists = resolve_multi(interpreter, reqs_to_resolve, platforms, None)
+
+  # FIXME: there's no description of what this does or why it's necessary.
+  overrides = {
+    'purelib': site_dir,
+    'headers': os.path.join(site_dir, 'headers'),
+    'scripts': os.path.join(site_dir, 'bin'),
+    'platlib': site_dir,
+    'data': site_dir
+  }
+
+  # The `python_dist` target builds for the current platform only.
+  # FIXME: why does it build for the current platform only?
```

Comment:
This was a cheap catch-all for preventing errors when the closure has platform-specific sources or dependencies.

Reply:
Could you describe this a bit more? Why are we indexing with `['current']` here? Or is there more info in #5825 somewhere?